File: python/day1.py
inp = []

with open("../input/input01.txt") as f:
    inp = [int(line) for line in f]

def part1(xs):
    return sum(xs)

def part2(xs):
    i = 0
    psum = 0
    seen = set()

    while True:
        psum += xs[i % len(xs)]
        i += 1        
        if psum in seen:
            return psum
        seen.add(psum)

print part1(inp)
print part2(inp)

File: haskell/Day1.hs
module Main (main) where

import AOC.Input
import AOC.Parsing

import qualified Data.Set as Set

year = 2018
day  = 1

parser :: Parser Int
parser = number

firstRecurring :: Ord a => [a] -> Maybe a
firstRecurring xs = search xs (Set.empty)
  where
    search []     _       = Nothing
    search (x:xs) seen
      | Set.member x seen = Just x
      | otherwise         = search xs (Set.insert x seen)

part1, part2 :: [Int] -> Int
part1 = sum
part2 xs = s
  where
    Just s = firstRecurring sums
    sums = scanl (+) 0 (cycle xs)

main :: IO ()
main = do
  input <- getParsedLines year day parser
  print $ part1 input
  print $ part2 input
